src/reconcile/rec_visibility.c

```diff
@@ -587,10 +630,15 @@ __wt_rec_upd_select(WT_SESSION_IMPL *session, WT_RECONCILE *r, WT_INSERT *ins, W
         supd_restore = F_ISSET(r, WT_REC_EVICT) &&
           (has_newer_updates || F_ISSET(S2C(session), WT_CONN_IN_MEMORY) ||
             page->type == WT_PAGE_COL_FIX);
-        WT_ERR(__rec_update_save(session, r, ins, rip,
-          upd_select->upd != NULL && upd_select->upd->type == WT_UPDATE_TOMBSTONE ? NULL :
-                                                                                    upd_select->upd,
-          supd_restore, upd_memsize));
+
+        onpage_upd = upd_select->upd != NULL && upd_select->upd->type == WT_UPDATE_TOMBSTONE ?
+          NULL :
+          upd_select->upd;
+
+        if (onpage_upd != NULL)
+            WT_ERR(__rec_validate_upd_chain(r, onpage_upd, select_tw, vpack));
+
+        WT_ERR(__rec_update_save(session, r, ins, rip, onpage_upd, supd_restore, upd_memsize));
         /*
          * Mark the selected update (and potentially the tombstone preceding it) as being destined
          * for the data store. Subsequent reconciliations should know that they can select this
```

Comment:
I'm going to leave the setting of WT_UPDATE_DS here as we should avoid any of the previous reasons that we were thinking of setting it in `__wt_hs_insert_updates`.